### DISCUSSION TITLE
Create company automatically from contact

### DISCUSSION THREAD

**Discussion by FelixMalfait** [twentyhq/twenty#4954] [Ideas]
When we create a contact, if that contact doesn't have a company, then we should create the company automatically from the domain name.

1. Create a function to detect if an email is a professional email (@company.com) or not (@gmail.com, @outlook.com etc)
You can use: https://gist.github.com/ammarshah/f5c2624d767f91a7cbdc4e54db8dd0bf
2. When a contact is created without a company. Check if it's a professional email and if a company exist with the contact email's domain, if yes then link the contact to that company.
3. If no company exist, and it's a professional email, then create company from domain name. Company name should be inferred from the domain name as v1 (remove domain extension + add a capital letter). In v2 we'll use a data enrichment database.

**Comment by Kanav-Arora** [comment]
Hey @FelixMalfait 
Can I work on this?

**Comment by Kanav-Arora** [comment]
There's an issue what if the input email is invalid. Is there a check for that?

**Comment by Bonapara** [comment]
Should this be part of a Workflows V1? We could create a workflow tab in settings to display the automation (non editable)

**Comment by FelixMalfait** [comment]
@Bonapara probably when we build workflow we can migrate it there yes. But I feel like this is a relatively simple task to implement VS implementing workflows = months of work. 
I tried to import our leads via CSV but it's a huge pain to have to create the companies manually since there's no way to create a relationship on the csv import. If the link with the company is created, then we can at least import a second csv to update the record in companies table. Building out a mechanism to handle relationships import/creation in the csv module seems like a heavy task, so the issue above offers a workaround. 

@Kanav-Arora thanks a lot. Is the issue clear enough? I think we already handle invalid email input. This is purely on the backend side.

**Comment by Kanav-Arora** [comment]
@FelixMalfait yeah
Its clear, I'll start working on it

**Comment by karishma-shah** [comment]
Hi @FelixMalfait , can I work on this please?

**Comment by Kanav-Arora** [comment]
> Hi @FelixMalfait , can I work on this please?

I have some pending issues to cover, you can work on this one. I haven't started it yet

**Comment by FelixMalfait** [comment]
@karishma-shah this one my be a bit hard for a first issue (Kanav had some experience with other issues priori to his). You can look for issues with the tag "good first issue". I see you've just done that today on already 🙏.

There are many different ways we could go to implement this, and I didn't give enough context in the initial ticket.
We don't have real-time sync between backend and frontend data, so triggering the enrichment action on a behavior like "escaping the email field" fields a bit hacky. Instead, we can start by implementing this as an action in the context menus (right click / object selected).

We have a moved to a new backend that creates layer of abstractions in the way we manage objects.
This task ranges from relatively easy to hard depending on how much we want to fit into doing the abstraction framework VS doing "if/else/hardcoded" implementations.
I'd say we could start abstracting the layout (frontend) but not abstract the backend.
In that case, here's how I would probably implement this:

### Backend
 **Enrichment endpoint** 
 Create an a enrichPerson mutation that takes the object id as input.
This function will be able to do various type of enrichment in the future. For example, find the email if there is only a name and company. But let's just focus on implementing a single use-case for now: if the person has an email, and that email is a "work email" (not gmail, etc.), and that person doesn't have a company. Then this endpoint should create the company.

**Layout**
- Create a LayoutMenuItems column (json) on objectMetadata 
- In worskpace-sync-metadata/standard-objects/person.object-metadata.ts declare the MenuItem {label: Enrich, mutation: enrichPerson}  
- Make sure it's populated when metadata is synced

### Frontend
- Upon loading the app, when objects' metadata is fetched, make sure menuItems are fetched
- Modify contextMenu + ActionBar + Showpage to display MenuItems

But this would need an internal discussion with @charlesBochet and team first, because we would be setting the first stone of the "layout framework" which will be an important piece

**Comment by Kanav-Arora** [comment]
Hi @FelixMalfait 
Can you please update here once the layout has been decided?
@karishma-shah If you want to try this and need help, you can join discord server for discussions regarding this issue.

**Comment by karishma-shah** [comment]
Hi @Kanav-Arora @FelixMalfait, no worries! I think it would be best if i picked up an issue with 'good first issue' tag for now instead.

**Comment by FelixMalfait** [comment]
@Kanav-Arora thanks I just created a dedicated issue to make sure we have a plan on layout, you can follow it here https://github.com/twentyhq/twenty/issues/2910

**Comment by BOHEUS** [comment]
@FelixMalfait is it still relevant, if so, should it be added as an additional step in flow of enrichment app or created as a separate app?

**Comment by BOHEUS** [comment]
Closing as it's done in https://github.com/twentyhq/twenty/pull/18909
